### PR TITLE
[ESP32] Added some OTA events for the device

### LINKED ISSUE #21955: [ESP32] Device should emit the OTA state changes for its internal usage
#### Problem
* Device should emit the OTA state changes
* Following can be the useful states
    * OTA DownloadIn Progress
    * OTA Download Complete
    * OTA Download Failed
    * OTA Download Aborted
    * OTA Apply InProgress
    * OTA Apply Complete
    * OTA Apply Failed

 #### Proposed Solution
* Add the events and emit them so that these can be consumed by the application

### PR DESCRIPTION
**Pull request by shubhamdp**
#### Problem
* Right now no way for the OTARequestor to get the OTA events other than polling the `UpdateState` attribute.
* Fixes #21955

#### Change overview
* Added few OTA events that can be consumed by the device

#### Testing
* Verified with the logs added in `ota-requestor-app/esp32`

```
I (12917) chip[DL]: OTA image download in progress
I (49437) chip[DL]: OTA image download complete
I (49567) chip[DL]: OTA image apply in progress
I (49567) chip[DL]: OTA image apply complete
```

### CODE REVIEW

File: src/platform/ESP32/OTAImageProcessorImpl.cpp [hunk 1]
```diff
@@ -211,17 +232,20 @@ void OTAImageProcessorImpl::HandleProcessBlock(intptr_t context)
...
         ESP_LOGE(TAG, "esp_ota_set_boot_partition failed (%s)!", esp_err_to_name(err));
+        PostOTAStateChangeEvent(DeviceLayer::kOtaApplyFailed);
         return;
     }
     ESP_LOGI(TAG, "Applying, Boot partition set offset:0x%x", imageProcessor->mOTAUpdatePartition->address);
 
     // HandleApply is called after delayed action time seconds are elapsed, so it would be safe to schedule the restart
-    chip::DeviceLayer::SystemLayer().StartTimer(chip::System::Clock::Milliseconds32(2 * 1000), HandleRestart, nullptr);
+    DeviceLayer::SystemLayer().StartTimer(System::Clock::Milliseconds32(2 * 1000), HandleRestart, nullptr);
+    PostOTAStateChangeEvent(DeviceLayer::kOtaApplyComplete);
```
Comment (dhrishi) [inline src/platform/ESP32/OTAImageProcessorImpl.cpp:248]: Want to send the event before triggering restart?

File: src/platform/ESP32/OTAImageProcessorImpl.cpp [hunk 2]
```diff
@@ -237,7 +261,7 @@ CHIP_ERROR OTAImageProcessorImpl::SetBlock(ByteSpan & block)
         {
             ReleaseBlock();
         }
-        uint8_t * mBlock_ptr = static_cast<uint8_t *>(chip::Platform::MemoryAlloc(block.size()));
```
Comment (dhrishi) [inline src/platform/ESP32/OTAImageProcessorImpl.cpp:240]: Why are these changes done?

File: src/include/platform/CHIPDeviceEvent.h [hunk 3]
```diff
@@ -287,6 +287,13 @@ enum ActivityChange
 enum OtaState
 {
     kOtaSpaceAvailable = 0,
+    kOtaDownloadInProgress,
+    kOtaDownloadComplete,
+    kOtaDownloadFailed,
+    kOtaDownloadAborted,
+    kOtaApplyInProgress,
+    kOtaApplyComplete,
+    kOtaApplyFailed,
```
Comment (dhrishi) [inline src/include/platform/CHIPDeviceEvent.h:296]: A one-liner description in front of each of these will help

File: src/platform/ESP32/OTAImageProcessorImpl.cpp [hunk 4]
```diff
@@ -37,11 +38,25 @@ void HandleRestart(Layer * systemLayer, void * appState)
...
     esp_restart();
 }
+
+void PostOTAStateChangeEvent(DeviceLayer::OtaState newState)
+{
+    DeviceLayer::ChipDeviceEvent otaChange;
+    otaChange.Type                     = DeviceLayer::DeviceEventType::kOtaStateChanged;
+    otaChange.OtaStateChanged.newState = newState;
+    CHIP_ERROR error                   = DeviceLayer::PlatformMgr().PostEvent(&otaChange);
+
```
Comment (dhrishi) [inline src/platform/ESP32/OTAImageProcessorImpl.cpp:48]: None of the OTA events require any kind of status/data to be sent?

File: src/platform/ESP32/OTAImageProcessorImpl.cpp (hunk 2 above)
Comment (shubhamdp) [reply src/platform/ESP32/OTAImageProcessorImpl.cpp:240]: This code is inside namespace chip, so using `chip::` is not required so, cleaning up.

File: src/include/platform/CHIPDeviceEvent.h (hunk 3 above)
Comment (shubhamdp) [reply src/include/platform/CHIPDeviceEvent.h:317]: Added the description.

File: src/platform/ESP32/OTAImageProcessorImpl.cpp (hunk 1 above)
Comment (shubhamdp) [reply src/platform/ESP32/OTAImageProcessorImpl.cpp]: Moved before starting the restart timer

File: src/platform/ESP32/OTAImageProcessorImpl.cpp (hunk 4 above)
Comment (shubhamdp) [reply src/platform/ESP32/OTAImageProcessorImpl.cpp:48]: These are simple state changes event and may not require additional data.

**Review by woody-apple — APPROVED** [review]

**Pull request merged**